### ISSUE TITLE
ColorDetector: needs re-calibration

### ISSUE THREAD

**Issue by myjeeed** [myjeeed/MakerWear#58]
There are currently two problems with this sensor:
1. It seems that it is ignoring the pull-down resistor, and will turn on when it is not connected to any other modules (i.e. the Power module)
2. The module requires to re-calibrate the colors. Right now it doesn't always detect colors perfectly.

This is the current table on the color means:

```
  byte black[3] = {85, 85, 65};
  byte blue[3] = {45, 75, 120};
  byte cyan[3] = {55, 130, 135};
  byte lime[3] = {60, 120, 60};
  byte yellow[3] = {255, 255, 100};
  byte red[3] = {165, 55, 60};
  byte magenta[3] = {130, 55, 100};
  byte white[3] = {255, 255, 255};
```

These values are the average values when I used the sensor on different colors.

**Comment by jonfroehlich** [comment]
Are you just doing a euclidean distance calculation to find closest
matching color?

Could likely do something more sophisticated here. I suggest talking to Lee.

Jon

On Tue, Jun 28, 2016 at 4:27 PM, Majeed notifications@github.com wrote:

> There are currently two problems with this sensor:
> 1. It seems that it is ignoring the pull-down resistor, and will turn on
> when it is not connected to any other modules (i.e. the Power module)
> 2. The module requires to re-calibrate the colors. Right now it doesn't
> always detect colors perfectly.
> 
> This is the current table on the color means:
> 
>   byte black[3] = {85, 85, 65};
>   byte blue[3] = {45, 75, 120};
>   byte cyan[3] = {55, 130, 135};
>   byte lime[3] = {60, 120, 60};
>   byte yellow[3] = {255, 255, 100};
>   byte red[3] = {165, 55, 60};
>   byte magenta[3] = {130, 55, 100};
>   byte white[3] = {255, 255, 255};
> 
> These values are the average values when I used the sensor on different
> colors.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> https://github.com/myjeeed/MakerWear/issues/58, or mute the thread
> https://github.com/notifications/unsubscribe/ABi-9ZgV-2e5rd-6tNBA-RRw1IBAyDakks5qQYPMgaJpZM4JAgJZ
> .

## 

Jon Froehlich
Assistant Professor
Computer Science
University of Maryland, College Park
http://www.cs.umd.edu/~jonf/
@jonfroehlich https://twitter.com/jonfroehlich - Twitter

**Comment by myjeeed** [comment]
Yes it's currently using euclidean distance (similar to how K-Means does its predictions)

Sure, just e-mailed Lee about this.

**Comment by myjeeed** [comment]
Lee:
You might do a bit better by converting to HSV or some other color representation (e.g. LAB) first and then only comparing the hue channel(s) with a circular distance (unless you need to support black/white or unsaturated colors, in which case you may need to use the other channels).

Also, colors are very subjective so unless you're always using consistent lighting and distance you're probably going to see a wide variation in readings. You could try doing expectation maximization (e.g., fitting a gaussian mixture model) around a set of known color readings, and then you'll have some meaningful probabilities for each of the color possibilities. Then you could do things like not output a prediction unless the confidence is above some threshold.

---

I think converting to HSV is a good idea, I should try it out.

**Comment by jonfroehlich** [comment]
Did Lee understand that we are working with a very small number of colors
in a highly discretized space? His email makes it seem like we're trying to
do perfect color matching...

Jon

On Mon, Jul 4, 2016 at 12:01 PM, Majeed notifications@github.com wrote:

> Lee:
> You might do a bit better by converting to HSV or some other color
> representation (e.g. LAB) first and then only comparing the hue channel(s)
> with a circular distance (unless you need to support black/white or
> unsaturated colors, in which case you may need to use the other channels).
> 
> Also, colors are very subjective so unless you're always using consistent
> lighting and distance you're probably going to see a wide variation in
> readings. You could try doing expectation maximization (e.g., fitting a
> gaussian mixture model) around a set of known color readings, and then
> you'll have some meaningful probabilities for each of the color
> possibilities. Then you could do things like not output a prediction unless
> 
> ## the confidence is above some threshold.
> 
> I think converting to HSV is a good idea, I should try it out.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/myjeeed/MakerWear/issues/58#issuecomment-230319651,
> or mute the thread
> https://github.com/notifications/unsubscribe/ABi-9bm6JcMFOOL9dzlfH7Z-Q4_51BW3ks5qSS5qgaJpZM4JAgJZ
> .

## 

Jon Froehlich
Assistant Professor
Computer Science
University of Maryland, College Park
http://www.cs.umd.edu/~jonf/
@jonfroehlich https://twitter.com/jonfroehlich - Twitter
